Add IPython rich display hook to AGraph class

dabreese00 opened the pull request:
Addresses feature request #182 .

I can't think of a way to add this method to the test suite, I guess it's too trivial for unit testing.  But if I should try let me know.

A way to manually confirm the functionality:

1. Install `pip install jupyterlab` and open Jupyter in browser (run `jupyter-lab`).
2. Create/open a new Notebook.
3. Run the following code in the Notebook:
    
    ```python
    import pygraphviz as pgv
    d = {"1": {"2": None}, "2": {"1": None, "3": None}, "3": {"2": None}}
    A = pgv.AGraph(d)
    A.layout()
    A
    ```

4.  The SVG image of the graph displays automatically.

dabreese00 committed:
Add IPython rich display hook to AGraph class

rossbar commented on pygraphviz/agraph.py:
```diff
@@ -230,6 +230,9 @@ def __repr__(self):
             return f"<AGraph {self.handle}>"
         return f"<AGraph {name} {self.handle}>"
 
+    def _repr_svg_(self):
+        return self.draw(format='svg').decode(self.encoding)
+
```
As it stands, this will give an `AttributeError` when users try to display an `AGraph` instance in a jupyter notebook cell when `layout` is called, which is a regression that has the potential to break a lot of existing notebooks.

I can think of a couple options: one would be to catch the `AttributeError` and return the original `repr` when the graph hasn't yet had its layout computed, e.g.:

```python
    def _repr_svg_(self):
        try:
            return self.draw(format='svg').decode(self.encoding)
        except AttributeError:
            return repr(self)
```

Another possible solution is to compute the layout inside `_repr_svg` in the case where it hasn't already been done. The biggest downside there is that the `layout` method might be computationally expensive depending on the method used and the size of the graph, which means executing a cell for a seemingly simple repr might result in unexpected behavior for users.

I think I'm partial to the first solution since it's the least invasive and avoids any major assumptions about what the user intended, but I'm curious to hear other opinions!

mjpieters replied:
Another option would be to implement [`_repr_mimebundle_()`](https://ipython.readthedocs.io/en/stable/config/integrating.html#MyObject._repr_mimebundle_), and switch representations based on whether or not layout has been called yet; include `image/svg+xml` if layout has been called, `text/plain` when not?

dabreese00 replied:
@rossbar Thanks for catching this.  @mjpieters I like that idea.  I also like the two ideas of @rossbar, I just hadn't got a chance to parse out which I preferred yet.

Now that I see the `_repr_mimebundle_()` suggestion, I think I like that the best, personally.  It seems to me this would allow the user to get the desired behavior -- whether text repr or layout + SVG repr -- in the most explicit/intuitive way, while not breaking the existing use cases.

I'd be happy to work at implementing that (or either of the others, whatever is decided to be the best).  To determine if `layout()` has been called or not, would it be sufficient to check `self.has_layout`?

rossbar reviewed (commented):
Thanks @dabreese00 for the proposal!

I think we still have to decide how best to implement this feature - see my comments below - but +1 for getting something like this in.

Note also that the CI failures are not related to your changes.

jarrodmillman commented:
@dabreese00 (Thanks for your contribution!)  I've made a few changes to the CI tests on the main branch.  Would you mind rebasing (or merging) on main?  If you want, I am more than happy to do it for you.  You will also need to run, black to format your changes automatically (so you don't have to worry about formatting the code).

You may want to briefly look at how sympy and pandas do this.  For example, in https://github.com/sympy/sympy/pull/20601 they add tests to check that the repr is what they expect (in this case `None`).  Here are more examples https://github.com/pandas-dev/pandas/blob/master/pandas/tests/frame/test_repr_info.py (from pandas).

dabreese00 commented:
@jarrodmillman Thanks!  I went ahead and tried rebasing for now.  I'll finish looking over those projects' approaches and get the `repr_mimebundle` method drafted, hopefully later this week, thanks for the links.  I'm a little new to this so if I did anything wrong please let me know.

MridulS commented:
I like the mimebundle approach too!

We could do something like
``` python
def _svg_repr(self):
    return self.draw(format="svg").decode(self.encoding)

def _repr_mimebundle_(self, include, exclude):
    if self.has_layout:
        repr_dict = {"image/svg+xml": self._svg_repr()}
    else:
        repr_dict = {"text/plain": self.__repr__()}
    return repr_dict
```
This should keep `_repr_mimebundle_` flexible enough that we can add new things without breaking backwards compatibility, maybe a LaTeX `repr` too one day! 🙃

Some ideas fo the tests too :)
``` python
# Some basic tests
def test_mimebundle():
    A = pgv.AGraph(name="test graph")
    A.add_path([1, 2, 3, 4])
    assert repr(A).startswith("<AGraph <Swig Object of type 'Agraph_t")
    A.layout()
    assert A._svg_repr().startswith("'<?xml version=")

def test_svg_repr_error(self):
    with pytest.raises(AttributeError):
        A = pgv.AGraph()
        A._svg_repr()

```

@dabreese00 What do you think?

dabreese00 commented:
@MridulS That looks excellent to me!  Thanks for a full implementation!  I will see if I can't get it submitted and passing tests (and run black) this weekend.  You guys are really making this easy for me!

dabreese00 committed:
Fix exception on IPython display without layout()

This fixes a regression introduced by the previous commit, where
using IPython to display an AGraph that had no layout would raise an
AttributeError.

With this fix, if layout() has not been called, IPython will just
display a basic text __repr__ of the AGraph; but if layout() has been
called, IPython can display an SVG image.

dabreese00 commented:
OK, I've pushed a fix.  It's almost a straight copy of what @MridulS wrote, I couldn't think of a lot to change or add, just one more test -- thanks again for that!  And thanks again to everyone helping out here.

If I should change anything, and/or squash the commits into one for a cleaner/less verbose history, let me know, happy to.

I guess the latter test, `test_svg_repr_error`, actually still passes when the `_svg_repr` method doesn't even exist (because that's also an AttributeError).  But the only ways I can think of to try to fix this obscure false passing result, seem more complex than they're worth.  So I probably won't address it, unless someone thinks otherwise.

jarrodmillman commented:
@dabreese00 @mjpieters Thanks!
